// glg-jwt/config/config.js
/*eslint no-void: "off"*/
"use strict";

/********************************************************************
 * Configuration
 ********************************************************************/

/*eslint no-process-env: "off"*/
module.exports = {
  enableLogging: typeof process.env.ENABLE_LOGGING !== "undefined",
  enableDebug: typeof process.env.ENABLE_DEBUG !== "undefined",
  enableVerbose: typeof process.env.ENABLE_VERBOSE !== "undefined",
  cssPath: `/${process.env.ORDERS_NAME}/css`,
  jsPath: `/${process.env.ORDERS_NAME}/js`,
  imgPath: `/${process.env.ORDERS_NAME}/img`,
  authServer: `/${process.env.ORDERS_NAME}/sendAuthEmail`,
  jwtSecret: process.env.JWT_SECRET,
  jwtExpirationInSeconds: process.env.JWT_EXPIRATION_IN_SECONDS,
  epiAuthTemplate: process.env.EPI_AUTH_TEMPLATE,
  jwtRedirectCookieName: process.env.JWT_REDIRECT_COOKIE_NAME || "jwturl",
  serverPort: process.env.PORT || 3000,
  authenticToken: process.env.STARPHLEET_AUTHENTIC_TOKEN,
  descriptions: {
    enableLogging: "Enable Basic Logging",
    enableDebug: "Enable Debug Logging",
    enableVerbose: "Enable Verbose Logging",
    serverPort: "Port For Express Server [Default: 3000]"
  }
};
